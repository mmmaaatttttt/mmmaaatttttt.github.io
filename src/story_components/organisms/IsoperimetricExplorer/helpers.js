import { mod } from "utils/mathHelpers";

// given an array of points and a point index,
// determine whether any segments joining the points together
// have a nontrivial intersection
function crossingExists(points, idx) {
  const segments = points.map((currentPt, i) => {
    const nextPtIdx = mod(i + 1, points.length);
    const nextPt = points[nextPtIdx];
    return {
      x1: currentPt.x,
      y1: currentPt.y,
      x2: nextPt.x,
      y2: nextPt.y
    };
  });

  // if an intersection exists, it must be with a segment whose point is at index idx
  const potentialCrossingSegments = [
    segments[idx],
    segments[mod(idx - 1, segments.length)]
  ];

  for (let segIdx in segments) {
    const seg = segments[segIdx];
    // skip over the candidate segments
    if (potentialCrossingSegments.includes(seg)) continue;

    // if you find an intersection, return true
    if (
      potentialCrossingSegments.some(otherSeg =>
        _doLinesIntersect(seg, otherSeg)
      )
    ) {
      return true;
    }
  }

  // if we make it out of the loop, there's no intersection
  return false;
}

// checks whether two segments (objects with x1, x2, y1, y2) intersect.
// cf: https://stackoverflow.com/questions/9043805/test-if-two-lines-intersect-javascript-function
function _doLinesIntersect(seg1, seg2) {
  const seg1XDiff = seg1.x2 - seg1.x1;
  const seg1YDiff = seg1.y2 - seg1.y1;
  const seg2XDiff = seg2.x2 - seg2.x1;
  const seg2YDiff = seg2.y2 - seg2.y1;
  const det = seg1XDiff * seg2YDiff - seg2XDiff * seg1YDiff;

  if (det === 0) return false;

  const crossXDiff = seg2.x2 - seg1.x1;
  const crossYDiff = seg2.y2 - seg1.y1;

  const lambda = (seg2YDiff * crossXDiff - seg2XDiff * crossYDiff) / det;
  const gamma = (seg1XDiff * crossYDiff - seg1YDiff * crossXDiff) / det;
  const lowerBound = 1e-6;
  const upperBound = 1 - lowerBound;
  const lambdaWithinBounds = lowerBound < lambda && lambda < upperBound;
  const gammaWithinBounds = lowerBound < gamma && gamma < upperBound;

  return lambdaWithinBounds && gammaWithinBounds;
}

export { crossingExists };
