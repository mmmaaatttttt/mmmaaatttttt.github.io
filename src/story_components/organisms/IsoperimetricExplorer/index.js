import React, { Component } from "react";
import PropTypes from "prop-types";
import {
  ClippedSVG,
  InteractivePolygon,
  LabeledSlider,
  StyledTable
} from "story_components";
import { average, mod, euclideanDistance } from "utils/mathHelpers";
import COLORS from "utils/styles";

class IsoperimetricExplorer extends Component {
  state = {
    points: [],
    baseArea: 1
  }

  componentDidMount() {
    this.generatePointsFromCount(this.props.initialSides);
  }

  crossingExists = (newPoint, idx) => {
    let { strokeWidth } = this.props;
    let { points } = this.state;
    let seg1 = {
      start: points[mod(idx - 1, points.length)],
      end: newPoint
    };
    let seg2 = {
      start: newPoint,
      end: points[mod(idx + 1, points.length)]
    };

    for (let i = 0; i < points.length; i++) {
      // two ways for there to be a problem:
      // 1. the newPoint intersects a non-adjacent line segment on the polygon
      // 2. an adjacent segment to newPoint intersects a point on the polygon

      if (i === idx) continue;
      let curPt = points[i];
      let nextIdx = mod(i + 1, points.length);
      let prevIdx = mod(i - 1, points.length);
      let nextPt = points[nextIdx];

      if (nextIdx !== idx) {
        let distanceFromNewPointToCurSeg = this.distanceBetween(curPt, nextPt, newPoint);
        if (distanceFromNewPointToCurSeg < strokeWidth) return true;

        let distanceFromSeg1ToCurPt = this.distanceBetween(seg1.start, seg1.end, curPt);
        if (distanceFromSeg1ToCurPt < strokeWidth) return true;
      }

      if (idx !== prevIdx) {
        let distanceFromSeg2ToCurPt = this.distanceBetween(seg2.start, seg2.end, curPt);
        if (distanceFromSeg2ToCurPt < strokeWidth) return true;
      }
    }

    return false;
  };

  generatePointsFromCount = newCount => {
    let points = Array.from({ length: newCount }, (_, i) => {
      const { width, height, initialSides } = this.props;
      const angle = (2 * Math.PI * i) / newCount - Math.PI / 2;

      // normalize the distance so that it's 100 when newCount is 3,
      // and otherwise is such that the area of the corresponding circle
      // is constant
      const distance = initialSides * 100 * Math.sin(Math.PI / initialSides) / 
        (newCount * Math.sin(Math.PI / newCount));
     
      return {
        x: width / 2 + distance * Math.cos(angle),
        y: height / 2 + distance * Math.sin(angle)
      };
    });
    
    this.setState({ points }, () => {
      // initialize base circle area
      // so numbers are easier to digest
      if (this.state.baseArea === 1) this.setBaseArea();
    });
  };

  getCenter = () => {
    let { points } = this.state;
    return {
      x: average(points, p => p.x),
      y: average(points, p => p.y)
    };
  };

  getLength = () => {
    let { points } = this.state;
    return points.reduce((length, point, i) => {
      let nextPoint = points[mod(i + 1, points.length)];
      let xSquare = (point.x - nextPoint.x) ** 2;
      let ySquare = (point.y - nextPoint.y) ** 2;
      let newDistance = (xSquare + ySquare) ** (1 / 2);
      return length + newDistance;
    }, 0);
  };

  handleDrag = (idx, coords) => {
    let points = [...this.state.points];
    let newPoint = { ...coords };
    if (!this.crossingExists(newPoint, idx)) {
      points[idx] = newPoint;
      this.setState({ points });
    }
  };

  handleValueChange = newVal => {
    this.generatePointsFromCount(newVal);
  };

  getCircleParams = () => {
    let center = this.getCenter();
    let r = this.getLength() / (2 * Math.PI);
    return { ...center, r };
  };

  getAreaInfo = radius => {
    let { points, baseArea } = this.state;
    let circleArea = Math.PI * radius ** 2;
    let normalizedCircleArea = 100 * circleArea / baseArea;

    // area calculation for arbitrary polygon, see:
    // https://www.mathopenref.com/coordpolygonarea2.html
    let polygonArea =
      Math.abs(points.reduce((area, point, i) => {
        let nextPoint = points[mod(i + 1, points.length)];
        return area + (point.x + nextPoint.x) * (-point.y + nextPoint.y);
      }, 0) / 2);
    let normalizedPolygonArea = 100 * polygonArea / baseArea;
    return {
      circleArea: normalizedCircleArea.toFixed(2),
      polygonArea: normalizedPolygonArea.toFixed(2),
      ratio: (polygonArea / circleArea).toFixed(2)
    };
  };

  distanceBetween = (lineStart, lineEnd, point) => {
    // a fun calculus problem
    // to calculate the distance between a segment
    // and a point not on that segment
    let x0 = lineStart.x;
    let x1 = lineEnd.x;
    let y0 = lineStart.y;
    let y1 = lineEnd.y;
    let p = point.x;
    let q = point.y;

    // the segment is parametrized by
    // (x0 + t(x0 - x1), y0 + t(y0 - y1)), 0 <= t <=1
    // using calculus, here's the value of t which minimizes distance to (p, q)
    let numerator = (x1 - x0) * (p - x0) + (y1 - y0) * (q - y0);
    let denominator = (x1 - x0) ** 2 + (y1 - y0) ** 2;
    let t0 = numerator / denominator;

    // distance from the point to the line at t0 is minimal,
    // unless t0 is outside of [0, 1].
    if (t0 < 0 || t0 > 1) {
      let d0 = euclideanDistance(x0 - p, y0 - q);
      let d1 = euclideanDistance(x1 - p, y1 - q);
      return Math.min(d0, d1);
    }

    let minD =
      Math.abs((x0 - p) * (y1 - y0) - (y0 - q) * (x1 - x0)) /
      denominator ** (1 / 2);
    return minD;
  };

  setBaseArea = () => {
    let { r } = this.getCircleParams();
    this.setState({ baseArea: Math.PI * r ** 2});
  }

  render() {
    const { width, height, initialSides, strokeWidth } = this.props;
    const { points } = this.state;
    let circleParams = this.getCircleParams();
    let { circleArea, polygonArea, ratio } = this.getAreaInfo(circleParams.r);
    let maxSides = 20;
    return (
      <div>
        <LabeledSlider
          min={initialSides}
          max={maxSides}
          step={1}
          value={points.length}
          title={`Number of district sides: ${points.length}`}
          handleValueChange={this.handleValueChange}
          color={COLORS.DARK_GRAY}
        />
        <ClippedSVG width={width} height={height} id="isoperimetric-svg">
          <circle
            cx={circleParams.x}
            cy={circleParams.y}
            r={circleParams.r}
            fill={COLORS.GRAY}
            stroke={COLORS.DARK_GRAY}
            strokeWidth={strokeWidth}
          />
          <InteractivePolygon
            points={points}
            handleDrag={this.handleDrag}
            strokeWidth={strokeWidth}
            fill={COLORS.GREEN}
            stroke={COLORS.DARK_GREEN}
          />
        </ClippedSVG>
        <StyledTable>
          <tbody>
            <td>Circle Area: {circleArea}</td>
            <td>Polygon Area: {polygonArea}</td>
            <td>Ratio: {ratio}</td>
          </tbody>
        </StyledTable>
      </div>
    );
  }
}

IsoperimetricExplorer.propTypes = {
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  initialSides: PropTypes.number.isRequired,
  strokeWidth: PropTypes.number.isRequired,
};

IsoperimetricExplorer.defaultProps = {
  width: 600,
  height: 400,
  initialSides: 3,
  strokeWidth: 3
};

export default IsoperimetricExplorer;
