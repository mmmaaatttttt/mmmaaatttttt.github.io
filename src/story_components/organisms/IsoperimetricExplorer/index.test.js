import React from "react";
import { render } from "@testing-library/react";
import IsoperimetricExplorer from ".";

describe("smoke and snapshot tests", () => {
  it("renders successfully", () => {
    render(<IsoperimetricExplorer />);
  });

  it("renders first demonstration successfully", () => {
    const { asFragment } = render(<IsoperimetricExplorer />);
    expect(asFragment()).toMatchSnapshot();
  });
});
