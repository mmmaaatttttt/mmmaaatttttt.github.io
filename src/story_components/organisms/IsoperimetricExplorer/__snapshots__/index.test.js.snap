// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smoke and snapshot tests renders first demonstration successfully 1`] = `
<DocumentFragment>
  <div>
    <div
      class="sc-bRBYWo gtDXeL"
    >
      <p
        class="sc-Rmtcm bImzes"
      >
        Number of district sides: 3
      </p>
      <section>
        <p
          class="sc-hzDkRC kxpnkL"
        >
          <i
            class="sc-gZMcBi dPNFn fas fa-minus fa-1x false"
            disabled=""
            style="color: rgb(85, 85, 85); opacity: 1;"
          />
        </p>
        <div
          class="rc-slider sc-chPdSV hlbFsG"
        >
          <div
            class="rc-slider-rail"
          />
          <div
            class="rc-slider-track"
            style="left: 0%; width: 0%;"
          />
          <div
            class="rc-slider-step"
          />
          <div
            aria-disabled="false"
            aria-valuemax="20"
            aria-valuemin="3"
            aria-valuenow="3"
            class="rc-slider-handle"
            role="slider"
            style="left: 0%; transform: translateX(-50%);"
            tabindex="0"
          />
          <div
            class="rc-slider-mark"
          />
          <div
            class="sc-iwsKbI gQUqIM"
            width="auto"
          >
            <div
              class="sc-jAaTju kaBuqs"
              color="#888"
              offset="10"
              width="6"
            />
            <div
              class="sc-jAaTju kbZoje"
              color="#e9e9e9"
              offset="10"
              width="6"
            />
          </div>
        </div>
        <p
          class="sc-hzDkRC kxpnkL"
        >
          <i
            class="sc-gZMcBi dPNFn fas fa-plus fa-1x false"
            disabled=""
            style="color: rgb(85, 85, 85); opacity: 1;"
          />
        </p>
      </section>
    </div>
    <div>
      <svg
        class="sc-gzVnrw kAIloo"
        viewBox="0 0 600 400"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <clippath
            id="clip-path-isoperimetric-svg"
          >
            <rect
              height="400"
              width="600"
              x="0"
              y="0"
            />
          </clippath>
        </defs>
        <g
          clip-path="url(#clip-path-isoperimetric-svg)"
        >
          <rect
            fill="none"
            height="400"
            stroke="#BBBBBB"
            stroke-width="3"
            width="600"
            x="0"
            y="0"
          />
          <circle
            cx="300"
            cy="200"
            fill="#BBBBBB"
            r="82.6993343132688"
            stroke="#555555"
            stroke-width="3"
          />
          <g>
            <polygon
              fill="#52A081"
              points="300,100 386.6025403784439,250 213.39745962155615,250.00000000000003"
              stroke="none"
              stroke-width="3"
            />
            <line
              stroke="#00802b"
              stroke-width="3"
              x1="300"
              x2="386.6025403784439"
              y1="100"
              y2="250"
            />
            <line
              stroke="#00802b"
              stroke-width="3"
              x1="386.6025403784439"
              x2="213.39745962155615"
              y1="250"
              y2="250.00000000000003"
            />
            <line
              stroke="#00802b"
              stroke-width="3"
              x1="213.39745962155615"
              x2="300"
              y1="250.00000000000003"
              y2="100"
            />
            <circle
              class="sc-jTzLTM fPdFLV react-draggable"
              cx="0"
              cy="0"
              fill="#52A081"
              r="8"
              stroke="#00802b"
              stroke-width="3"
              style=""
              transform="translate(300,100)"
            />
            <circle
              class="sc-jTzLTM fPdFLV react-draggable"
              cx="0"
              cy="0"
              fill="#52A081"
              r="8"
              stroke="#00802b"
              stroke-width="3"
              style=""
              transform="translate(386.6025403784439,250)"
            />
            <circle
              class="sc-jTzLTM fPdFLV react-draggable"
              cx="0"
              cy="0"
              fill="#52A081"
              r="8"
              stroke="#00802b"
              stroke-width="3"
              style=""
              transform="translate(213.39745962155615,250.00000000000003)"
            />
          </g>
        </g>
      </svg>
    </div>
    <table
      class="sc-kgoBCf kChdgN"
    >
      <tbody>
        <tr>
          <td>
            Circle Area: 100.00
          </td>
          <td>
            Polygon Area: 60.46
          </td>
          <td>
            Ratio: 0.60
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
